R/calcHDDCDD.R

```diff
@@ -1,101 +1,714 @@
-              tail(seq(1, 0, length.out = length(convergenceYears)), -1),
-              rep(0, length(afterConvergenceYears)))
-  names(lambda) <- c(yearsHistory, yearsScenario)
-  lambda <- as.magpie(lambda)
-
-  rcps     <- grep("history", getNames(data, TRUE)$rcp,      value = TRUE, invert = TRUE)
-  ssps     <- grep("history", getNames(data, TRUE)$scenario, value = TRUE, invert = TRUE)
-  variable <- grep("history", getNames(data, TRUE)$variable, value = TRUE, invert = TRUE)
-
-  data[is.na(data)] <- 0
-  data <- mbind(data, (1 / 3) * (setYears(data[, 2099, ], 2100) +
-                                 setYears(data[, 2098, ], 2100) +
-                                 setYears(data[, 2097, ], 2100))) # give a point for 2100
-
-  # distribute the history between all scenarios
-  data <- do.call("mbind", lapply(rcps, function(rcp) {
-    do.call("mbind", lapply(ssps, function(ssp) {
-      tmp <- data[, , rcp][, , ssp]
-      tmp[, yearsHistory, ] <- setNames(data[, yearsHistory, "history"][, , "history"],
-                                        getNames(tmp[, yearsHistory, ]))
+#' @author Robin Krekeler, Hagen Tockhorn
+#' @examples
+#'
+#' \dontrun{
+#' calcHDDCDD()
+#' }
+#'
+#' @importFrom madrat getConfig
+#' @importFrom raster cellStats
+#' @importFrom ncdf4 nc_open
+#' @importFrom tidyr %>%
+#' @importFrom dplyr mutate
+#' @importFrom rlang .data
+#' @importFrom pracma integral2
+#' @importFrom stringr str_sub
+#' @importFrom terra app nlyr tapp subset rast classify time
+
+
+calcHDDCDD <- function(mappingFile, bait=FALSE, multiscen = FALSE) {
+
+
+  # FUNCTIONS-------------------------------------------------------------------
+
+  # fill dates for unnamed data
+  #   ---> was supposed to go to convertISIMIP, doesn't work with country check
+  fillDates <- function(r, filename, bait=FALSE) {
+    if (grepl(".nc|.nc4", filename)) {
+      filename <- gsub(".nc|.nc4", "", filename)
+    }
+
+    if (bait) {
+      dStart <- as.Date(stringr::str_sub(filename, -17, -10),
+                        format = "%Y%m%d")
+      n <- nlyr(r)
+      dates <- seq.Date(dStart, by = "day", length.out = n)
+    }
+    else {
+      yStart <- stringr::str_sub(filename, -9, -6)
+      dStart <- as.Date(paste0(yStart, "-1-1"))
+      n <- nlyr(r)
+
+      dates <- seq.Date(dStart, by = "day", length.out = n)
+    }
+
+    # fill dates
+    names(r) <- dates
+    return(r)
+  }
+
+
+  # prepare input for calcBAIT()
+  calcBaitInput <- function(frsds=NULL, fsfc=NULL, fhuss=NULL, baitInput=NULL, mean=FALSE) {
+    if(mean) {
+      # optional: calculate daily means over years to fill missing data
+      baitInputMean <- sapply(
+        names(baitInput), function(var) {
+          mean <- tapp(baitInput[[var]],
+                              unique(substr(names(baitInput[[var]]), 6, 11)),
+                              fun = "mean")
+          names(mean) <- gsub("\\.", "-", substr(names(mean), 2, 6))
+          return(mean)})
```

Comment:
This double-use of variable names confused me. `meanVar`
 is just a suggestion. There is probably something more indicative.
```suggestion
          meanVar <- tapp(baitInput[[var]],
                              unique(substr(names(baitInput[[var]]), 6, 11)),
                              fun = "mean")
          names(meanVar) <- gsub("\\.", "-", substr(names(meanVar), 2, 6))
          return(meanVar)})
```